Fix broken #copy_from method in Budget class

sempervictus opened the pull request:
This addresses [#34966](https://community.openproject.com/projects/openproject/work_packages/34966)

Commit:
Fix broken call to #super in Budget#copy_from

The #copy_from method in Budget appears to assume that superclasses
will have the same method available to them as called on the
new Budget object to which the copy is going. This is not the case.

Resolve the issue by assigning duplicated attributes from the
source object directly to self.attributes prior to copying the
actual budget items.

Tested in Chef/Vagrant test environment.
Deployed to SemperVictus pre-production OP instance successfully.

sempervictus commented:
So here's a weird artifact - it looks like the _copied_ object **takes away** the original object's assigned planned costs on copy instead of copying references to them.

sempervictus commented:
There's something really unpleasant going on under the skin - some sort of hooks causing deletion in what _should be ephemeral rails console operations_ - 
```
irb(main):003:0> n = Budget.new
  AnonymousUser Load (1.8ms)  SELECT "users".* FROM "users" WHERE "users"."type" = $1 ORDER BY "users"."id" ASC LIMIT $2  [["type", "AnonymousUser"], ["LIMIT", 1]]
=> #<Budget id: nil, project_id: nil, author_id: 3, subject: nil, description: nil, fixed_date: nil, created_at: nil, updated_at: nil>
irb(main):004:0> n.copy_from budget
  LaborBudgetItem Load (0.8ms)  SELECT "labor_budget_items".* FROM "labor_budget_items" WHERE "labor_budget_items"."budget_id" = $1 ORDER BY labor_budget_items.id ASC  [["budget_id", 1]]
  User Load (1.2ms)  SELECT "users".* FROM "users" WHERE "users"."type" IN ($1, $2, $3, $4, $5) AND "users"."id" = $6  [["type", "User"], ["type", "AnonymousUser"], ["type", "DeletedUser"], ["type", "SystemUser"], ["type", "Users::InexistentUser"], ["id", 4]]
  LaborBudgetItem Load (0.6ms)  SELECT "labor_budget_items".* FROM "labor_budget_items" WHERE "labor_budget_items"."budget_id" = $1 ORDER BY labor_budget_items.id ASC  [["budget_id", 1]]
  User Load (0.7ms)  SELECT "users".* FROM "users" WHERE "users"."type" IN ($1, $2, $3, $4, $5) AND "users"."id" = $6  [["type", "User"], ["type", "AnonymousUser"], ["type", "DeletedUser"], ["type", "SystemUser"], ["type", "Users::InexistentUser"], ["id", 4]]
   (0.4ms)  BEGIN
  LaborBudgetItem Destroy (0.6ms)  DELETE FROM "labor_budget_items" WHERE "labor_budget_items"."id" = $1  [["id", 1]]
   (46.3ms)  COMMIT
  MaterialBudgetItem Load (0.8ms)  SELECT "material_budget_items".* FROM "material_budget_items" WHERE "material_budget_items"."budget_id" = $1 ORDER BY material_budget_items.id ASC  [["budget_id", 1]]
  CostType Load (0.9ms)  SELECT "cost_types".* FROM "cost_types" WHERE "cost_types"."id" IN ($1, $2, $3, $4, $5, $6, $7, $8)  [["id", 1], ["id", 3], ["id", 2], ["id", 4], ["id", 5], ["id", 6], ["id", 7], ["id", 8]]
  MaterialBudgetItem Load (0.7ms)  SELECT "material_budget_items".* FROM "material_budget_items" WHERE "material_budget_items"."budget_id" = $1 ORDER BY material_budget_items.id ASC  [["budget_id", 1]]
  CostType Load (0.6ms)  SELECT "cost_types".* FROM "cost_types" WHERE "cost_types"."id" IN ($1, $2, $3, $4, $5, $6, $7, $8)  [["id", 1], ["id", 3], ["id", 2], ["id", 4], ["id", 5], ["id", 6], ["id", 7], ["id", 8]]
   (0.4ms)  BEGIN
  MaterialBudgetItem Destroy (0.6ms)  DELETE FROM "material_budget_items" WHERE "material_budget_items"."id" = $1  [["id", 1]]
  MaterialBudgetItem Destroy (0.5ms)  DELETE FROM "material_budget_items" WHERE "material_budget_items"."id" = $1  [["id", 2]]
  MaterialBudgetItem Destroy (0.5ms)  DELETE FROM "material_budget_items" WHERE "material_budget_items"."id" = $1  [["id", 3]]
  MaterialBudgetItem Destroy (0.5ms)  DELETE FROM "material_budget_items" WHERE "material_budget_items"."id" = $1  [["id", 4]]
  MaterialBudgetItem Destroy (0.7ms)  DELETE FROM "material_budget_items" WHERE "material_budget_items"."id" = $1  [["id", 5]]
  MaterialBudgetItem Destroy (0.6ms)  DELETE FROM "material_budget_items" WHERE "material_budget_items"."id" = $1  [["id", 6]]
  MaterialBudgetItem Destroy (0.6ms)  DELETE FROM "material_budget_items" WHERE "material_budget_items"."id" = $1  [["id", 7]]
  MaterialBudgetItem Destroy (0.5ms)  DELETE FROM "material_budget_items" WHERE "material_budget_items"."id" = $1  [["id", 8]]
  MaterialBudgetItem Destroy (0.5ms)  DELETE FROM "material_budget_items" WHERE "material_budget_items"."id" = $1  [["id", 9]]
```
Even trying to poke at this in `rails c` is a destructive operation...

Commit:
Resolve budget child item copy confusion

The #copy_from method does not account for child models duplicated
from another Budget holding a unique constraint validation on the
:budget_id attribute. The Budget itself also appears to inherit
the source Budget's ID, creating additional problems.

Address by unfolding the sub-object copy procedure and clean up
the relevant attributes on #dup ensuring the appropriate relations
are created when the object is assigned (after the #dup block).

Tested in Chef/Vagrant workflow.

ulferts commented on modules/budgets/app/models/budget.rb:
```diff
@@ -76,10 +76,19 @@ def initialize(attributes = nil)
+        d.budget = self
+        d.budget_id = self.id
+      end
+    end
```
I believe this can be streamlined a bit which should also make the code easier to understand:
```
def self.new_copy(source)
  source = (source.is_a?(Budget) ? source : self.class.find(source))
  copy = new(source.attributes.slice('project_id', 'subject', 'description', 'fixed_date').merge(author => User.current))
  source.labor_budget_items.each do |bi|
    copy.labor_budget_items.build(bi.attributes.slice('units', 'cost_type_id', 'comments', 'amount'))
  end
  source.material_budget_items.each do |bi|
    copy.material_budget_items.build(bi.attributes.slice('units', 'cost_type_id', 'comments', 'amount'))
  end
end
```

There are a couple of changes in here:
1) it is turned into a class method as I believe this method will not be fit to merge data from two budgets but it can create a copy of an existing budget. The call in the controller will have to be adapted.
2) It assigns the current user to be the author of the budget. This I believe to be important as not to suggest that somebody created a budget when it was copied by a different user.

ulferts reviewed (changes requested):
Hi @sempervictus,

any thank you for the PR. I have suggested a different implementation in code which I believe should be easier to understand. 

Can you please double check that it still works?

sempervictus commented:
Doesn't work, ID seems to fail (i replaced the content of #copy_from with your #new_copy):
```
existentUser"], ["status", 1], ["id", 2], ["LIMIT", 1]]
D, [2020-11-05T15:43:49.046385 #59651] DEBUG -- :    (0.9ms)  SELECT MAX("settings"."updated_on") FROM "settings"
D, [2020-11-05T15:43:49.058754 #59651] DEBUG -- :   Budget Load (1.2ms)  SELECT "budgets".* FROM "budgets" WHERE "budgets"."id" = $1 LIMIT $2  [["id", 5], ["LIMIT", 1]]
D, [2020-11-05T15:43:49.081425 #59651] DEBUG -- :   Project Load (1.3ms)  SELECT "projects".* FROM "projects" WHERE "projects"."id" = $1  [["id", 3]]
D, [2020-11-05T15:43:49.094720 #59651] DEBUG -- :   User Load (0.5ms)  SELECT "users".* FROM "users" WHERE "users"."type" IN ($1, $2, $3, $4, $5) AND "users"."id" = $6  [["type", "User"], ["type", "AnonymousUser"], ["type", "SystemUser"], ["type", "DeletedUser"], ["type", "Users::InexistentUser"], ["id", 2]]
D, [2020-11-05T15:43:49.098823 #59651] DEBUG -- :    (0.4ms)  SELECT "enabled_modules"."name" FROM "enabled_modules" WHERE "enabled_modules"."project_id" = $1  [["project_id", 3]]
D, [2020-11-05T15:43:49.100671 #59651] DEBUG -- :   Budget Load (0.3ms)  SELECT "budgets".* FROM "budgets" WHERE "budgets"."id" = $1 LIMIT $2  [["id", 5], ["LIMIT", 1]]
D, [2020-11-05T15:43:49.113408 #59651] DEBUG -- :    (0.5ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
E, [2020-11-05T15:43:49.115591 #59651] ERROR -- : [current_user=OpenProject Admin] Couldn't find Budget without an ID: Couldn't find Budget without an ID
D, [2020-11-05T15:43:49.129207 #59651] DEBUG -- :   CustomStyle Load (0.8ms)  SELECT "custom_styles".* FROM "custom_styles" ORDER BY created_at DESC LIMIT $1  [["LIMIT", 1]]
D, [2020-11-05T15:43:49.147539 #59651] DEBUG -- :    (0.7ms)  SELECT MAX(union_query.max_updated_at)
FROM ((SELECT MAX(updated_at) AS max_updated_at FROM statuses) UNION (SELECT MAX(updated_at) AS max_updated_at FROM enumerations) UNION (SELECT MAX(updated_at) AS max_updated_at FROM types))
AS union_query

D, [2020-11-05T15:43:49.164747 #59651] DEBUG -- :    (0.3ms)  SELECT "schema_migrations"."version" FROM "schema_migrations" ORDER BY "schema_migrations"."version" ASC
D, [2020-11-05T15:43:49.181235 #59651] DEBUG -- :   Attachment Load (0.5ms)  SELECT "attachments".* FROM "attachments" WHERE "attachments"."container_id" = $1 AND "attachments"."container_type" = $2 AND "attachments"."description" = $3 ORDER BY attachments.created_at LIMIT $4  [["container_id", 2], ["container_type", "Principal"], ["description", "avatar"], ["LIMIT", 1]]
I, [2020-11-05T15:43:49.184192 #59651]  INFO -- : method=GET path=/budgets/5/copy format=html controller=BudgetsController action=copy status=500 duration=228.73 view=57.73 db=45.94 user=2
```

sempervictus commented:
Noticed your comment regarding the rewrite to a class method and controller changes. Sounds great, what i have there so far works to implement the functionality for now, and i'm sure one of the paid developers can re-implement when they have time in a manner better suited to readability and long term maintainability of the paid product.
There are other broken pieces in the release (like managed repos being inoperable) which i'll need to address from a functional perspective when i get time to do so before i can allocate resources to update the controller and class method adaptation for OP's long term use.

Commit:
Rewrite for readability

DRY the code, leaving it as an instance method for the time being.

TODO for upstream:
  Convert to class method and update controller in the MVC (per
GH comment).
  Write tests/spec and verify results of the original fix and final
revision for release branches.

ulferts commented:
Closing in favour of #8856 which extends the changes made in here.
